### PR TITLE
feat(IPX): add PageView in IPX actions history

### PR DESCRIPTION
**Pull request by dguerinCoveo**
Allows the tracking of pageViews in the actions history for IPX. 

image

My first plan was to use coveo.analytics package to modify the actions.history array in the web local storage directly from the atomic component. However, coveo.analytics is imported in browser mode in the atomic package when I install it, and this mode [does not contain the history utils](https://github.com/coveo/coveo.analytics.js/blob/master/src/coveoua/browser.ts#L29). 

Therefore, I need to have a headless action to update the actions history array in local storage, so here we are.

### CODE REVIEW

File: packages/headless/src/features/analytics/generic-analytics-actions-loader.ts [hunk 1]
```diff
@@ -44,6 +46,15 @@ export interface GenericAnalyticsActionCreators {
...
   logCustomEvent(payload: LogCustomEventActionCreatorPayload): CustomAction;
+
+  /**
+   * Adds an entry in the action history.
+   * @param itemPermanentId - The unique identifier of the item.
+   * @returns A dispatchable action.
+   */
+  addPageViewEntryInActionsHistory(
+    itemPermanentId: string
+  ): AsyncThunkAction<void, string, {}>;
```
Comment (dguerinCoveo) [inline packages/headless/src/features/analytics/generic-analytics-actions-loader.ts:57]: How do you feel about adding that there @louis-bompart ? Is there another place which would make more sense?

File: packages/headless/src/index.ts [hunk 2]
```diff
@@ -550,6 +550,7 @@ export {buildResultTemplatesManager} from './features/result-templates/result-te
 export * from './features/analytics/search-analytics-actions-loader';
 export * from './features/analytics/click-analytics-actions-loader';
 export * from './features/analytics/generic-analytics-actions-loader';
+export * from './features/actions-history/ipx-actions-history-actions-loader';
```
Comment (dguerinCoveo) [inline packages/headless/src/index.ts:553]: I feel like this goes here and not in the region grouped section, but I'm guessing more than anything

Comment (olamothe) [reply packages/headless/src/index.ts:553]: The problem is the IPX naming: There's no "IPX" package in Headless like there is for commerce, case assist, insight panel. It's confusing.

I would export it as part of classic/generic analytics functionality.

ie: 
* Delete `features/actions-history/ipx-actions-history-actions-loader.ts`
* Delete `features/actions-history/ipx-actions-history-actions.ts`
* Add the action in `features/analytics/analytics-action.ts`
* Export it in `features/analytics/generic-analytics-action-loader.ts`

File: packages/headless/src/features/actions-history/ipx-actions-history-actions.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    const state = getState() as SearchAppState;
+    if (state.configuration.analytics.enabled) {
+      historyStore.addElement({
+        name: 'PageView',
+        value: itemPermanentId,
+        time: JSON.stringify(new Date()),
+      });
+    }
+  }
+);
```
Comment (olamothe) [inline packages/headless/src/features/actions-history/ipx-actions-history-actions.ts:17]: Coupled with with other comment below, I would make this just a generic function for all entry in action history, not something just specific for pageview VS IPX.

```ts
export const addEntryInActionsHistory = createAsyncThunk(
  'analytics/actionsHistory/add',
  async (entry: HistoryElement, {getState}) => {
    const state = getState() as SearchAppState;
    if (state.configuration.analytics.enabled) {
      historyStore.addElement(entry);
    }
  }
);
```

File: packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  public connectedCallback() {
+    const slotName = 'attributes';
+    this.hasDefaultSlot = !!getDefaultSlotFromHost(this.host);
+    this.linkAttributes = getAttributesFromLinkSlot(this.host, slotName);
+  }
+
+  public async onSelect() {
+    const resultPermanentId = this.result.raw.permanentid;
+    if (resultPermanentId) {
+      const action = loadIPXActionsHistoryActions(
```
Comment (olamothe) [inline packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx:80]: I would load the action on `initialize`.

**Review by olamothe — CHANGES_REQUESTED** [review]
It seems we don't have any IPX package in Headless -- Which is a bummer.

File: packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx (hunk 4 above)
Comment (dguerinCoveo) [reply packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx:80]: good point

File: packages/atomic/src/components.d.ts [hunk 5]
```diff
@@ -5229,6 +5248,15 @@ declare namespace LocalJSX {
...
         "collapseFacetsAfter"?: number;
     }
+    /**
+     * The `atomic-ipx-result-link` component automatically transforms a search result title into a clickable link that points to the original item. It is an experimental internal component not intended for general use.
+     */
+    interface AtomicIpxResultLink {
+        /**
+          * Specifies a template literal from which to generate the `href` attribute value (see [Template literals](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals)).  The template literal can reference any number of result properties from the parent result. It can also reference the window object.  For example, the following markup generates an `href` value such as `http://uri.com?id=itemTitle`, using the result's `clickUri` and `itemtitle` fields. ```html <atomic-ipx-result-link href-template='${clickUri}?id=${raw.itemtitle}'></atomic-ipx-result-link> ```
+         */
+        "hrefTemplate"?: string;
```
Comment (mikegron) [inline packages/atomic/src/components.d.ts:5258]: tu a combiné les modifs pour le href template?

File: packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ * @slot default - Lets you display alternative content inside the link
+ * @slot attributes - Lets you pass [attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#attributes) down to the link element, overriding other attributes, to be used exclusively with an "a" tag such as `<a slot="attributes" target="_blank" download></a>`.
+ * @internal
+ */
+@Component({
+  tag: 'atomic-ipx-result-link',
+  styleUrl: 'atomic-ipx-result-link.pcss',
+  shadow: false,
+})
+export class AtomicIPXResultLink implements InitializableComponent {
```
Comment (mikegron) [inline packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx:37]: Ont peux tu hérité de result link direct?

**Review by mikegron — APPROVED** [review]

File: packages/atomic/src/components.d.ts (hunk 5 above)
Comment (dguerinCoveo) [reply packages/atomic/src/components.d.ts:5258]: Non, AtomicResultLink definit deja une prop pour hrefTemplate. C'est juste du copy pasta

File: packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx (hunk 6 above)
Comment (louis-bompart) [reply packages/atomic/src/components/ipx/atomic-ipx-result-link/atomic-ipx-result-link.tsx:37]: No Inheritance for components is our rule of thumb. Do prefer composition :)

**Comment by dguerinCoveo** [comment]
@olamothe, @louis-bompart gentle reminder to review the PR please :)

**Review by louis-bompart — CHANGES_REQUESTED** [review]
I have to concur with Olivier, please create an IPX package to contain and isolate the code unique to IPX.

**Comment by dguerinCoveo** [comment]
> I have to concur with Olivier, please create an IPX package to contain and isolate the code unique to IPX.

In headless you mean? I think Olivier's point was the opposite. When he reviewed, I had the headless actions inside IPX specific packages. I moved my headless code to the generic package following his comment.

**Review by olamothe — APPROVED** [review]

**Comment by olamothe** [comment]
> I have to concur with Olivier, please create an IPX package to contain and isolate the code unique to IPX.

For Headless specifically (and for the analytics actions added in this PR), I don't see the need to create an IPX package, or IPX specific files. Pushing entries in actions-history (for legacy events) is a "generic" event that can be used by anyone, even for classic search pages.

**Review by louis-bompart — CHANGES_REQUESTED** [review]
After discussion with @olamothe:
 - Indeed, an IPX entrypoint at this stage is overkill, let's not.
 - Instead, let's go back to the solution originally proposed:
   - Dedicated action-loader for IPX
 The rationale is that PageView in ActionHistory is an out-of-standard feature, so we do not want that to be included by our users (and we won't support it either).
 Moreover, all this will be wiped in V3.
 
 TL;DR: `git reset --hard HEAD~1` and it should be alright.

**Comment by dguerinCoveo** [comment]
Changed the package to the reached consensus. I've re-asked only for your approval @louis-bompart , but tell me if we need oli's eyes on that 🙏

**Review by louis-bompart — APPROVED** [review]

**Review by pathiery — APPROVED** [review]
